Emby.Server.Implementations/SyncPlay/SyncPlayManager.cs

```diff
@@ -41,338 +44,415 @@ public class SyncPlayManager : ISyncPlayManager, IDisposable
-            if (!IsSessionInGroup(session))
-            {
-                return;
+                    AddSessionToGroup(session, group);
+                    group.CreateGroup(session, request, cancellationToken);
+                }
             }
-
-            LeaveGroup(session, CancellationToken.None);
         }
 
-        private bool IsSessionInGroup(SessionInfo session)
-        {
-            return _sessionToGroupMap.ContainsKey(session.Id);
-        }
-
-        private bool HasAccessToItem(User user, Guid itemId)
+        /// <inheritdoc />
+        public void JoinGroup(SessionInfo session, JoinGroupRequest request, CancellationToken cancellationToken)
         {
-            var item = _libraryManager.GetItemById(itemId);
-
-            // Check ParentalRating access
-            var hasParentalRatingAccess = !user.MaxParentalAgeRating.HasValue
-                || item.InheritedParentalRatingValue <= user.MaxParentalAgeRating;
-
-            if (!user.HasPermission(PermissionKind.EnableAllFolders) && hasParentalRatingAccess)
+            if (!IsRequestValid(session, request))
             {
-                var collections = _libraryManager.GetCollectionFolders(item).Select(
-                    folder => folder.Id.ToString("N", CultureInfo.InvariantCulture));
-
-                return collections.Intersect(user.GetPreference(PreferenceKind.EnabledFolders)).Any();
+                return;
             }
 
-            return hasParentalRatingAccess;
-        }
-
-        private Guid? GetSessionGroup(SessionInfo session)
-        {
-            _sessionToGroupMap.TryGetValue(session.Id, out var group);
-            return group?.GetGroupId();
-        }
-
-        /// <inheritdoc />
-        public void NewGroup(SessionInfo session, CancellationToken cancellationToken)
-        {
             var user = _userManager.GetUserById(session.UserId);
 
-            if (user.SyncPlayAccess != SyncPlayAccess.CreateAndJoinGroups)
+            // Locking required to access list of groups.
+            lock (_groupsLock)
             {
-                _logger.LogWarning("NewGroup: {0} does not have permission to create groups.", session.Id);
+                _groups.TryGetValue(request.GroupId, out IGroupController group);
 
-                var error = new GroupUpdate<string>
+                if (group == null)
                 {
-                    Type = GroupUpdateType.CreateGroupDenied
-                };
+                    _logger.LogWarning("Session {SessionId} tried to join group {GroupId} that does not exist.", session.Id, request.GroupId);
 
-                _sessionManager.SendSyncPlayGroupUpdate(session.Id, error, CancellationToken.None);
-                return;
-            }
+                    var error = new GroupUpdate<string>(Guid.Empty, GroupUpdateType.GroupDoesNotExist, string.Empty);
+                    _sessionManager.SendSyncPlayGroupUpdate(session, error, CancellationToken.None);
```

Comment:
I don't like that the syncplay commands are part of the sessionmanager. I'd rather the sessionmanager exposed a generic method for sending commands.

Not a huge deal right now though, but food for thought.

Reply:
I totally agree.